extendr-macro.qmd

```diff
@@ -0,0 +1,376 @@
+
+::: callout-note
+In essence, all items that are returned from a function must be able to be turned into an `Robj`. Other extendr types such as `List`, for example, have a `From<T> for Robj` implementation that defines how it is converted into an `Robj`.
+:::
+
+This means that with a little extra work, the `Shape` enum can be returned to R. To do so, the `#[extendr]` macro needs to be added to an impl block.  
+
+
+## Exporting `impl` blocks
+
+The other supported item that can be made available to R is an [`impl`](https://doc.rust-lang.org/std/keyword.impl.html) block. 
+`impl` is a keyword that allows you to _implement_ a trait or an inherent implementation. The `#[extendr]` macro works with inherent implementations. These are `impl`s on a type such as an `enum` or a `struct`. extendr _does not_ support using `#[extendr]` on trait impls. 
+
+::: callout-note
+You can only add an inherent implementation on a type that you have own and not provided by a third party crate. This would violate the [orphan rules](https://github.com/Ixrec/rust-orphan-rules?tab=readme-ov-file#what-are-the-orphan-rules).
+:::
+
+Continuing with the `Shape` example, this enum alone cannot be returned to R. For example, the following code will result in a compilation error
+
+```rust
+#[derive(Debug)]
+enum Shape {
+    Triangle,
+    Rectangle,
+    Pentagon,
+    Hexagon,
+}
+
+#[extendr]
+fn make_shape(shape: &str) -> Shape {
+    match shape {
+        "triangle" => Shape::Triangle,
+        "rectangle" => Shape::Rectangle,
+        "pentagon" => Shape::Pentagon,
+        "hexagon" => Shape::Hexagon,
+        &_ => unimplemented!()
+    }
+}
+```
+```
+error[E0277]: the trait bound `Shape: ToVectorValue` is not satisfied
+  --> src/lib.rs:19:1
+   |
+19 | #[extendr]
+   | ^^^^^^^^^^ the trait `ToVectorValue` is not implemented for `Shape`, which is required by `extendr_api::Robj: From<Shape>`
+   |
+```
+
+However, if an impl block is added to the `Shape` enum, it can be returned to R. 
+
+```{extendrsrc}
+#[derive(Debug)]
+enum Shape {
+    Triangle,
+    Rectangle,
+    Pentagon,
+    Hexagon,
+}
+
+#[extendr]
+impl Shape {
+    fn new(x: &str) -> Self {
+        match x {
+            "triangle" => Self::Triangle,
+            "rectangle" => Self::Rectangle,
+            "pentagon" => Self::Pentagon,
+            "hexagon" => Self::Hexagon,
+            &_ => unimplemented!(),
+        }
+    }
+
+    fn n_coords(&self) -> usize {
+        match &self {
+            Shape::Triangle => 3,
+            Shape::Rectangle => 4,
+            Shape::Pentagon => 4,
+            Shape::Hexagon => 5,
+        }
+    }
+}
+```
+
+In this example two new methods are added to the `Shape` enum. The first `new()` is like the `make_shape()` function that was shown earlier: it takes a `&str` and returns an enum variant. Now that the enum has an `impl` block with `#[extendr]` atrtibute macro, it can be exported to R by inclusion in the `extendr_module! {}` macro.
```

Comment:
```suggestion
In this example two new methods are added to the `Shape` enum. The first `new()` is like the `make_shape()` function that was shown earlier: it takes a `&str` and returns an enum variant. Now that the enum has an `impl` block with `#[extendr]` attribute macro, it can be exported to R by inclusion in the `extendr_module! {}` macro.
```